sys/include/entropy_source/entropy_source.h

```diff
@@ -0,0 +1,289 @@
+/**
+ * @brief   Data structure for Adaptive Proportion Test (NIST SP 800-90B 4.4.2).
+ */
+typedef struct {
+    uint8_t old_sample;     /**< Preceding sample to compare for repetition */
+    uint16_t cnt_prop;      /**< Counter to count proportion */
+    uint16_t cnt_window;    /**< Counter to count window size */
+    uint16_t c_prop;        /**< Cutoff threshold */
+} entropy_source_tests_prop_t;
+
+/**
+ * @brief   Scale Min. Entropy to fixed point integer to avoid float. The entropy per sample
+            (8 Byte) of a noise source can likely be smaller than 1 bit.
+ */
+#define ENTROPY_SOURCE_HMIN_SCALE(x)    ((unsigned)(x * (1 << 16)))
+
+/**
+ * @brief   Scale internal fixed point Min. Entropy back to float. This macro is not
+ *          required and only there for convenience.
+ */
+#define ENTROPY_SOURCE_HMIN_SCALE_BACK(x)    (((float)x / (1 << 16)))
+
+/**
+ * @brief   Window size for Adaptive Proportion Test (NIST SP 800-90B 4.4.2).
+ *
+ * In (NIST SP 800-90B 4.4.2) a window size of 512 samples is recommended for non-binary
+ * sources. Typically, RIOT use cases will not request as many samples, thus, it might be worth
+ * considering a smaller window size so the test is more likely to complete a cycle. It is
+ * noteworthy that a cutoff value calculated by @ref entropy_source_test_prop_cutoff that is
+ * greater than the window size may lead to undetected errors.
+ */
+#ifndef CONFIG_ENTROPY_SOURCE_TESTS_WIN
+#define CONFIG_ENTROPY_SOURCE_TESTS_WIN     (512)
+#endif
+
+/**
+ * @brief   Abort factor for von Neumann extractor. The algorithms runs as long as no bit
+ *          changes appear in subsequent samples. This define adds a factor that
+            aborts the procedure after (factor * requested length) samples.
+ */
+#ifndef CONFIG_ENTROPY_SOURCE_NEUMANN_ABORT
+#define CONFIG_ENTROPY_SOURCE_NEUMANN_ABORT (5)
+#endif
+
+/**
+ * @brief   Get one sample of the entropy source.
+ *
+ * This function is typically used by the entropy source internally. A conditioning
+ * component might need an interface to request a variable number of samples, e.g., depending
+ * on the contained amount of entropy.
+ *
+ * @param[in] sample  pointer to write sample to.
+ *
+ * @return            ENTROPY_SOURCE_OK on success
+ * @return            negative @ref entropy_source_error_t code on error
+ */
+typedef int (*entropy_source_sample_func_t)(uint8_t *sample);
+
+/**
+ * @brief   Applies von Neumann unbiasing.
+ *
+ * This function requests as many samples needed to create \p len unbiased bytes using
+ * \p func. The algorithm compares bits of consecutive samples. Only bit changes will be
+ * considered for the output value. An abort criterium stops sampling after
+ * (len * CONFIG_ENTROPY_SOURCE_NEUMANN_ABORT) iterations.
+ *
+ * @warning This function has a non-deterministic runtime.
+ *
+ * @param[in] func  pointer to @ref entropy_source_sample_func_t function that returns samples
+ * @param[in] out   pointer to write unbiased bytes to
```

Comment:
```suggestion
 * @param[out] out   pointer to write unbiased bytes to
```
Please check, there may be others